packages/flutter/lib/src/widgets/nested_scroll_view.dart

```diff
@@ -1367,7 +1368,19 @@ class _NestedScrollPosition extends ScrollPosition implements ScrollActivityDele
 
   @override
   void restoreScrollOffset() {
-    if (coordinator.canScrollBody)
+    // If the children of a [NestedScrollView] have [PageStorageKey], the
+    // scroll offset should be restored correctly and not controlled by
+    // [_NestedScrollCoordinator.canScrollBody].
+    Element? elementWithPageStorageKey;
+    context.storageContext.visitAncestorElements((Element element) {
+      if (element.widget.key is PageStorageKey) {
+        elementWithPageStorageKey = element;
+        return false;
+      }
+      return true;
+    });
+    final NestedScrollViewState? ancestorStateIsNestedScrollViewState = elementWithPageStorageKey?.findAncestorStateOfType<NestedScrollViewState>();
+    if (identical(ancestorStateIsNestedScrollViewState, coordinator._state) || coordinator.canScrollBody)
```

Comment:
I haven't investigated, but why's `coordinator.canScrollBody` false in the situations where we want to restore the scroll offset?

Reply:
This is because of the #14133 change due to #10658.

Reply:
I believe this test case is one of the reasons: https://github.com/flutter/flutter/issues/10658#issuecomment-350886639